xdsl/transforms/stencil_to_csl_stencil.py

```diff
@@ -171,10 +295,314 @@ def match_and_rewrite(self, op: dmp.SwapOp, rewriter: PatternRewriter, /):
             nested_rewriter.rewrite_op(new_apply_op)
 
 
+def get_op_split(
+    ops: Sequence[Operation], prefetch: BlockArgument
+) -> tuple[Sequence[Operation], Sequence[Operation]]:
+    """
+    Returns a split of ops into an `(a,b)` tuple, such that:
+
+    - `a` contains neighbour accesses plus the minumum set of instructions to reduce the accessed data to 1 thing
+    - `b` contains everything else
+
+    If no valid split can be found, return `(ops, [])`.
+
+    This function does not attempt to arithmetically re-structure the computation to obtain a good split. To do this,
+    `RestructureSymmetricReductionPattern()` may be executed first.
+    """
+    a: Sequence[Operation] = []
+    b: Sequence[Operation] = []
+    rem: Sequence[Operation] = []
+    for op in ops:
+        if isinstance(op, csl_stencil.AccessOp):
+            (b, a)[op.op == prefetch].append(op)
+        else:
+            rem.append(op)
+
+    has_changes = True
+    while (
+        len(
+            a_exports := set(
+                op
+                for op in a
+                for r in op.results
+                for u in r.uses
+                if u.operation not in a
+            )
+        )
+        > 1
+        and has_changes
+    ):
+        a_deps = set(
+            u.operation for op in a_exports for r in op.results for u in r.uses
+        )
+        has_changes = False
+        movable = [
+            op
+            for op in a_deps
+            if all(x.op in a for x in op.operands if isinstance(x, OpResult))
+        ]
+        if len(movable) > 0:
+            has_changes = True
+            for op in movable:
+                a.append(op)
+                rem.remove(op)
+
+    if len(a_exports) == 1:
+        return a, b + rem
+
+    # fallback
+    # always place `stencil.return` in second block?
+    return ops[:-1], [ops[-1]]
+
+
+@dataclass(frozen=True)
+class ConvertApplyOpPattern(RewritePattern):
+    """
+    Fuses a `csl_stencil.prefetch` and a `stencil.apply` to build a `csl_stencil.apply`.
+
+    If there are several candidate prefetch ops, the one with the largest result buffer size is selected.
+    The selection is greedy, and could in the future be expanded into a more global selection optimising for minimal
+    prefetch overhead across multiple apply ops.
+    """
```

Comment:
Nice! Really helpful already! 

Ideally you'd mention what parameters your pattern takes (e.g. `num_chunks`) so that people know how to call it!

Reply:
Ok